[DCJ-610] JobModel includes more info on targeted resource

okotsopoulos opened the pull request:
__Jira ticket__: https://broadworkbench.atlassian.net/browse/DCJ-610

## Addresses

When debugging flights on behalf of users, it's not always clear from a flight's description or log spelunking what resource is being targeted.  We do not want to have to connect directly to the database to confirm, and would like to avoid needless back-and-forth with users.

## Summary of changes

Expand `JobModel` response body to include more information on the resource targeted by the flight: `targetIamResource` with fields `type`, `id`, and `action`.

We already pass this information as flight inputs for the majority of our flights, to enable Sam-based permission checks when retrieving information on flights.

When this change is merged in, we'll benefit from this change for flights that have already run, since we have several years of history where we pass in these input parameters.

## Testing Strategy

Expanded unit tests.

Locally, added additional print statements to unit tests for a visual into what the beefed-up `JobModel`s will look like:
```
class JobModel {
    id: fgK5VbjTTkCXv9F633zCnA
    description: flight2
    jobStatus: succeeded
    statusCode: 418
    submitted: null
    completed: null
    className: bio.terra.service.job.JobServiceTestFlight
    targetIamResource: class JobTargetResourceModel {
        type: dataset
        id: 6a430aa8-a84d-4ac3-8ed3-19ca7cc17c57
        action: share_policy::reader
    }
}
```

<!-- Reminder -->
<!-- Two CODEOWNERS will be automatically assigned to review this pull request. If you otherwise have two reviewers, you do not need to wait for their review. -->

okotsopoulos committed:
[DCJ-610] JobModel includes more info on targeted resource

Specifically, iam_resource_type, iam_resource_id, and iam_resource_action.

Motivation: when debugging flights on behalf of users, it's not always clear from a flight's description or log spelunking what resource is being targeted.

We already pass this information as flight inputs for the majority of our flights, to enable Sam-based permission checks when retrieving information on flights.

pshapiro4broad commented on src/main/java/bio/terra/service/job/JobService.java:
```diff
@@ -282,14 +288,23 @@ public JobModel mapFlightStateToJobModel(FlightState flightState) {
+            .iamResourceId(iamResourceId);
+    if (iamResourceType != null) {
+      jobModel.iamResourceType(iamResourceType.getSamResourceName());
+    }
```
This is a case where I think `Optional.ofNullable()` can help. Using it avoids duplicating the expression you're checking for `!= null`. Also, I think it's clearer to use the non-fluent version of the setter for a non-fluent usage like this.
```suggestion
    Optional.ofNullable(iamResourceType)
      .map(IamResourceType::getSamResourceName)
      .forEach(jobModel::setIamResourceType);
```

okotsopoulos replied:
Regarding the Optional, this is a style that hasn't caught on consistently through the repo, and in this service the existing convention is to perform null checks.  Optional usage like this can at times make code more difficult for a wider audience to understand, in this case the duplication is worth it to me.

> Also, I think it's clearer to use the non-fluent version of the setter for a non-fluent usage like this.

I agree with you, a later commit made that change.

okotsopoulos committed:
Reformat target resource information in JobModel

It makes more sense to group the type, ID, and action together in a single JobTargetResourceModel object.

Expanded unit tests to exercise cases where these values are not available as input parameters.

rushtong reviewed (approved):
👍🏽

pshapiro4broad commented on src/test/java/bio/terra/service/job/JobServiceTest.java:
```diff
@@ -324,32 +316,106 @@ private Matcher<JobModel>[] getJobMatchers(List<JobModel> jobModels) {
 
-  // Submit a flight; wait for it to finish; return the flight id
+  /**
+   * Submit a flight with optional input parameters and wait for it to finish.
```
I think this is a copy/paste error, I don't see where this submits a flight. Same with `expectedJobModel(int i)` below

okotsopoulos replied:
It submits a flight via a call to `runFlight`, but it buries the call:
```
    return new JobModel()
        .id(runFlight(description, flightClass, testUser, IAM_RESOURCE_ID, shouldAddParameters))
        …
```

I will extract that call to make the behavior clearer.

pshapiro4broad replied:
Oh, no problem then! Maybe change the method name to better match its behavior too? That's what helped to fool me here

okotsopoulos replied:
Yes, good shout!  I renamed these methods (ex. `runFlightAndReturnExpectedJobModel`).

pshapiro4broad reviewed (approved):
Looks OK to me, just a minor comment about possibly using `Optional` and what looks like a copy/paste error the javadoc.

okotsopoulos committed:
test - rename test methods to clarify behavior
